Blaze: Connect product selector view to "Create" button in Campaign List

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11669 (wrote the number wrong in the branch name, sorry).
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR creates a navigation Coordinator `BlazeCampaignCreationCoordinator` to help decide entry into Blaze native campaign creation screen, depending on number of available products. The Coordinator is then used by three different entry points:
- campaign creation in Blaze dashboard view
- campaign creation in Blaze campaign list screen
- campaign creation in Product form

Additionally, this PR also updates the product selector to also include non-purchasable products, as we decided to follow the Blaze behavior on web which allows creating campaign for them. Related Slack conversation in: p1704712552381949-slack-C03L1NF1EA3

## Not included in this PR:

We want to show alert when user taps "Create" but there is no existing product. This is an edge case and requires a bit more coding:

> You currently don’t have a product available for promotion. Would you like to create a product now? [Cancel] [Create Product]

Slack p1705041492389469/1705033010.195009-slack-C03L1NF1EA3

It will be handled in a separate PR. Currently the button simply does nothing when tapped in this case.

This also does not include unit tests for `BlazeCampaignCreationCoordinator`. The PR is already big so I'll make it in a separate PR.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Multiple products path

1. Have a Blaze-eligible site with multiple products,
2. Go to Menu > Blaze,
3. Tap "Create",
4. Ensure product selector view is shown as a sheet,
5. Tap on a product,
6. Ensure the campaign creation form is shown.
7. Try the back button to test navigation.

### Single product path
1. Have a Blaze-eligible site with a single product,
2. Go to Menu > Blaze,
3. Tap "Create",
4. Ensure campaign creation form is shown as part of parent nav (not a sheet),
7. Try the back button to test navigation.

Also possible to test the above from Show All Campaigns on dashboard. The above should work the same.

Also possible to test with intro screen shown, everything should still work the same.

### Multiple products:
https://github.com/woocommerce/woocommerce-ios/assets/266376/ebcb2bf0-66b7-4b6d-8a91-f460eab8684d 

### Single product with intro and then without intro:

https://github.com/woocommerce/woocommerce-ios/assets/266376/c78c851e-6ff7-4a89-ac9e-3a5008059c55

hafizrahman committed:
Remove purchasable = true check in dashboard because we let non-purchasable to be advertised too.

This matches web behavior.

hafizrahman committed:
Check if there's multiple Blaze-eligible products in the site.

hafizrahman committed:
Refactor web campaign creation navigation to its own function.

hafizrahman committed:
Make product selector also show non-purchasable items when opened from Blaze dashboard.

hafizrahman committed:
Add navigation to product selector from "Create" button.

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Also pass product ID navigating from dashboard's product selector to campaign form, now that the form supports it.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan reviewed (commented):
Could you kindly resolve the conflicts? Thank you!

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardView.swift

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/180">16.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman commented:
Setting this back to draft to handle the case for "no eligible product available when tapping Create button".

hafizrahman committed:
Refactor create button navigation logic to be handled by the view model instead.

Also adds an error case but not yet handled.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented:
This might be semi-related but I found an issue when opening the campaign creation from the dashboard for a store with a single product. The campaign creation screen is presented correctly, but there is no button to dismiss the view (we can not drag to dismiss in landscape mode because it's full screen). Can you squeeze in a fix for that here?


https://github.com/woocommerce/woocommerce-ios/assets/5533851/feb0117e-0bac-457c-b61b-f7db2718dc0a

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Update unit test with new rule.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
If going to native campaign creation form directly without product selector, show() the form.

This way it becomes part of the current navigation stack and gets a back button automatically.

hafizrahman commented:
> The campaign creation screen is presented correctly, but there is no button to dismiss the view (we can not drag to dismiss in landscape mode because it's full screen). Can you squeeze in a fix for that here?

Added fix in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/272ea7f963306642e146c71f9908dbbe6f3e808c

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Add unit tests related to create campaign destination

hafizrahman commented:
@selanthiraiyan I fixed the conflict and updated the test instructions, this is good for review now.

hafizrahman committed:
Add issue number for todo.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/180">16.9</a>. The due date for this milestone has already passed.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListViewModel.swift:
```diff
@@ -11,6 +12,16 @@ extension BlazeCampaign: Identifiable {
+    enum CreateCampaignDestination: Equatable {
+        case productSelector
+        case campaignForm(productID: Int64)
+        case webviewForm(productID: Int64?)
```
Super nit: 
```suggestion
        case webViewForm(productID: Int64?)
```

hafizrahman replied:
This is handled in the new Coordinator class https://github.com/woocommerce/woocommerce-ios/pull/11671/files#diff-97e95c754013a21f67871677e78df49649e1cd07195b4f1a5f3128a19ac2516aR11

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListViewModel.swift:
```diff
@@ -146,10 +182,36 @@ private extension BlazeCampaignListViewModel {
+
+    // Figure out what view should be presented when "Create" button is tapped, depending on the number of eligible products.
+    func determineDestination() -> CreateCampaignDestination {
+        switch productResultsController.fetchedObjects.count {
+        case 0:
+            return .noProductAvailable
+        case 1:
+            guard let product = self.productResultsController.fetchedObjects.first else {
+                // Should not happen since we already checked the count, but just in case.
+                return .noProductAvailable
+            }
+            return .campaignForm(productID: product.productID)
+        default:
+            return .productSelector
+        }
+    }
```
What do you think about using if else if? This lets us use `isEmpty` on `fetchedObjects` instead of using hard-coded zero. 


```suggestion
        if productResultsController.fetchedObjects.isEmpty {
            return .noProductAvailable
        } else if productResultsController.fetchedObjects.count == 1,
                  let recentPublishedProduct = productResultsController.fetchedObjects.first {
            return .campaignForm(productID: recentPublishedProduct.productID)
        } else {
            return .productSelector
        }
```

hafizrahman replied:
I implemented this in the Coordinator https://github.com/woocommerce/woocommerce-ios/pull/11671/files#diff-97e95c754013a21f67871677e78df49649e1cd07195b4f1a5f3128a19ac2516aR95-R104

Thank you!

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListView.swift:
```diff
@@ -63,6 +57,81 @@ final class BlazeCampaignListHostingController: UIHostingController<BlazeCampaig
+            self?.handlePostCreation()
+        }
+        let webViewController = AuthenticatedWebViewController(viewModel: webViewModel)
+        self.navigationController?.show(webViewController, sender: self)
```
`self` not needed here.

hafizrahman replied:
This is removed completely as part of refactor, thanks!

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazeCampaignListViewModelTests.swift:
```diff
@@ -383,6 +383,99 @@ final class BlazeCampaignListViewModelTests: XCTestCase {
         XCTAssertEqual(viewModel.selectedCampaignURL?.absoluteString, "https://wordpress.com/advertising/campaigns/123/example.com?source=campaign_list")
     }
 
+    // MARK: - Create campaign navigation
```
Nice tests. ❤️

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardViewModelTests.swift:
```diff
@@ -452,7 +452,8 @@ final class BlazeCampaignDashboardViewModelTests: XCTestCase {
 
-    func test_shouldRedactView_is_true_after_reload_finishes_when_only_non_purchasable_product_available() async {
+    // Creating Blaze campaign for non-purchasable product is allowed.
+    func test_shouldRedactView_is_false_after_reload_finishes_when_only_non_purchasable_product_available() async {
```
IMO, we can remove this test as we have stopped checking for purchasable condition. As there is no code about purchasable condition it might cause unnecessary confusion. What do you think?

hafizrahman replied:
Good point, I agree. removed in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/3633f5afc56a33aa6e3fba55553d144f7e5d5ea5

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardView.swift:
```diff
@@ -93,11 +92,13 @@ private extension BlazeCampaignDashboardViewHostingController {
         } else {
-            blazeNavigationController.viewControllers = [controller]
-            parentNavigationController?.present(blazeNavigationController, animated: true)
+            self.navigationController?.show(controller, sender: self)
```
`self` not needed.

hafizrahman replied:
This is also removed completely as part of refactor, thanks!

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardView.swift:
```diff
@@ -49,7 +48,7 @@ private extension BlazeCampaignDashboardViewHostingController {
         let productSelectorViewModel = ProductSelectorViewModel(
             siteID: viewModel.siteID,
-            purchasableItemsOnly: true,
+            purchasableItemsOnly: false,
```
Can we skip sending a value for `purchasableItemsOnly` as `false` is the default value?

hafizrahman replied:
Done in the Controller, it no longer sends the value.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListViewModel.swift:
```diff
@@ -134,9 +163,16 @@ private extension BlazeCampaignListViewModel {
         resultsController.onDidResetContent = { [weak self] in
```
We should rename this into `blazeCampaignResultsController` as we now have another `ResultsController` in this file.

hafizrahman replied:
The productResultsController is refactored out of the viewmodel so this is resolved.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListViewModel.swift:
```diff
@@ -11,6 +12,16 @@ extension BlazeCampaign: Identifiable {
 
 /// View model for `BlazeCampaignListView`
 final class BlazeCampaignListViewModel: ObservableObject {
+    enum CreateCampaignDestination: Equatable {
```
Nice work adding an `enum` for this. 👏

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListView.swift:
```diff
@@ -63,6 +57,81 @@ final class BlazeCampaignListHostingController: UIHostingController<BlazeCampaig
+    /// Handles navigation to the Blaze product selector view
+    func navigateToBlazeProductSelector(source: BlazeSource) {
+        blazeNavigationController.viewControllers = [productSelectorViewController]
+        self.navigationController?.present(blazeNavigationController, animated: true, completion: nil)
```
`self` not needed.

hafizrahman replied:
This is removed completely as part of refactor.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListView.swift:
```diff
@@ -63,6 +57,81 @@ final class BlazeCampaignListHostingController: UIHostingController<BlazeCampaig
+        if blazeNavigationController.presentingViewController != nil {
+            blazeNavigationController.show(controller, sender: self)
+        } else {
+            self.navigationController?.show(controller, sender: self)
```
`self` is not needed.

hafizrahman replied:
This is also removed completely as part of refactor.

selanthiraiyan reviewed (commented):
Nice work, Hafiz! This works as expected. 👏 

I left a few comments about the code. 

Could you also check the possibility of moving the navigation code to a [coordinator](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Tools/Coordinator.swift) and reusing that between the dashboard, campaign list and product form screens?

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
First attempt at Blaze campaign creation navigation coordinator.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Refactor @MainActor usage in BlazeCampaignCreationFormViewModel.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Move destination logic into Coordinator.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Refactor BlazeCampaignList to use the Coordinator for navigation.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Refactor BlazeCampaignDashboard to navigate using the coordinator.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Remove potentially confusing unit test.

We always allow non-purchasable products to be Blazed now.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Remove unneeded unit tests. These will need to be in the Controller instead.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Make parentNavigationController in BlazeCampaignDashboard's hosting controller non-optional.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman commented:
@itsmeichigo This is now good for another round of review, thanks!

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,165 @@
+    private let source: BlazeSource
+    private let storageManager: StorageManagerType
+    private let featureFlagService: FeatureFlagService
+    var navigationController: UINavigationController
```
Please keep this as a constant with `let` unless you need to change it later.

hafizrahman replied:
This was just conforming to `Coordinator`.  TIL that it can be changed to `let`.

itsmeichigo replied:
This statement in the protocol: 

```Swift
var navigationController: UINavigationController { get }
```

Means that to conform the protocol, a struct/class needs to have a readonly property named `navigationController`. It doesn't define whether the property should be mutable within the scope of that struct/class. So you can decide whether to use a private(set) variable or a constant. In this specific case, the property doesn't need to be mutable so a constant is the best choice.

hafizrahman replied:
Awesome, thank you for the clear explanation.

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/commit/9f84b035a8632961e804653d90c6c7852dc97028

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,165 @@
+            return .noProductAvailable
+        } else if productResultsController.fetchedObjects.count == 1,
+                  let recentPublishedProduct = productResultsController.fetchedObjects.first {
+            return .campaignForm(productID: recentPublishedProduct.productID)
```
Nit: `productResultsController.fetchedObjects` is a calculated variable, so to avoid duplicated calculations, we should keep a reference to the calculated variable and use it for each check.

```Swift
let fetchedObjects = productResultsController.fetchedObjects
if fetchedObjects.isEmpty {
// TODO
} else if fetchedObjects.count == 1, let firstProduct = fetchedObject.first {
// TODO
} else {
// TODO
}
```

hafizrahman replied:
Handled as part of https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/9f84b035a8632961e804653d90c6c7852dc97028

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,165 @@
+    private func updateCreateCampaignDestination() {
+        blazeCreationEntryDestination = featureFlagService.isFeatureFlagEnabled(.blazei3NativeCampaignCreation)
+        ? determineDestination()
+        : .webViewForm(productID: productResultsController.fetchedObjects.first?.productID)
```
Nit: For readability, you can use an `if` check instead of the ternary operator.

hafizrahman replied:
Handled as part of https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/9f84b035a8632961e804653d90c6c7852dc97028

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,165 @@
+    /// Handles navigation to the native Blaze creation
+    func navigateToNativeCampaignCreation(source: BlazeSource, productID: Int64) {
+        let controller = BlazeCampaignCreationFormHostingController(
+            viewModel: .init(siteID: self.siteID,
```
Nit: We should move away from implicit init where possible for performance reasons. You can declare the view model in a separate statement and set it in the declaration of the controller.

hafizrahman replied:
Can you share a link explaining about the performance issue? I tried googling but couldn't find the right set of keywords.

I also see this being done quite a bit in the codebase. If this is an old pattern that we should avoid going forward, I wonder if we can set a rule around it.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/d805ebc4dd630bf706f53aedb988c474a82e8ab6

itsmeichigo replied:
I shared a link about this the other day here [p1704251212353159-slack-C02Q9GACF] - sorry for not including it in this comment.

hafizrahman replied:
Thanks! My bad for forgetting this, I knew I've read it somewhere 🤦🏼

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,165 @@
+    /// Handles navigation to the Blaze product selector view
+    func navigateToBlazeProductSelector(source: BlazeSource) {
+        // View controller for product selector before going to campaign creation form.
+        var productSelectorViewController: ProductSelectorViewController {
```
Nit: You can use a calculated constant here instead of variable.

```Swift
let controller: ProductSelectorViewController = {
// TODO
}()
```

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/275a1081b1ec7e1030ae05f8324053eb2f5be195

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardView.swift:
```diff
@@ -7,10 +7,11 @@ import Kingfisher
-    private let parentNavigationController: UINavigationController?
+    private let parentNavigationController: UINavigationController
     private lazy var blazeNavigationController = WooNavigationController()
+    private var coordinator: BlazeCampaignCreationCoordinator?
```
Nit: You can declare the coordinator lazily to avoid optional here.

```suggestion
    private lazy var coordinator = BlazeCampaignCreationCoordinator(
            siteID: viewModel.siteID,
            siteURL: viewModel.siteURL,
            source: .myStoreSection,
            navigationController: parentNavigationController,
            onCampaignCreated: handlePostCreation
        )
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardView.swift:
```diff
@@ -21,21 +22,34 @@ final class BlazeCampaignDashboardViewHostingController: SelfSizingHostingContro
+            }
+            self.parentNavigationController.dismiss(animated: true) {
+                coordinator.start()
             }
```
Please be careful with nested closures, it's best to reference `self` weakly to avoid any retain cycle:

```suggestion
            self?.parentNavigationController.dismiss(animated: true) {
                self?.coordinator.start()
            }
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Blaze/BlazeCampaignDashboardView.swift:
```diff
@@ -21,21 +22,34 @@ final class BlazeCampaignDashboardViewHostingController: SelfSizingHostingContro
+            navigationController: parentNavigationController,
+            onCampaignCreated: self.handlePostCreation
+        )
+        self.coordinator = coordinator
```
You can remove this if you declare the coordinator lazily above.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/c50165126169778c816a66c5a16715ad1d4088cb

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewController.swift:
```diff
@@ -777,9 +777,15 @@ extension DashboardViewController {
             removeBlazeCampaignView()
         }
+
+        guard let navController = navigationController else {
```
Nit: You can simplify this check:

```suggestion
        guard let navigationController else {
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazeCampaignListViewModelTests.swift:
```diff
@@ -456,4 +456,10 @@ private extension BlazeCampaignListViewModelTests {
         storage.saveIfNeeded()
     }
+
+    func insertProduct(_ readOnlyProduct: Product) {
+        let newProduct = storage.insertNewObject(ofType: StorageProduct.self)
+        newProduct.update(with: readOnlyProduct)
+        storage.saveIfNeeded()
+    }
```
Is this being used in this file?

hafizrahman replied:
Deleted in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/21223a1f937eb0e9334d8958975960f40220b445

This was missed when I deleted unit tests from this file. Reminder about unused functions / vars is something I wish Xcode would copy from Android Studio 😅

itsmeichigo reviewed (changes requested):
Thanks for the introduction of the new coordinator! I found a regression issue for the web flow. Steps to reproduce:

- Disable the feature flag `blazei3NativeCampaignCreation`.
- Build and run the app
- Log in to a store with at least one campaign created in the past.
- Tap Create campaign on the Dashboard screen.
- Notice that the web view shows the creation form for the latest product.

-> Expectation: the web should display the product list to select from.
-> Cause: You are using the last product to send to the web flow. I think you are doing the same with the native flow. 
-> Suggestion: You should inject an optional product ID to the coordinator. Then update the logic for creating the destination:

- If it's the web flow, send the optional ID as-is to the web view model. The rest is taken care of.
- If it's the native flow:
  - if the ID is present, open the campaign creation flow with it.
  - if not, check if there are multiple products in the store. If yes, show the product selector, else open the creation flow with the single product.

Which brings me to the second point: you should integrate the coordinator to the product form too. The product ID in that case will always be present.

Extra: it'd be nice to have unit tests for the coordinator to double check the logic. Please feel free to work on it in a separate PR.

hafizrahman committed:
Refactor controller logic to fix regression.

hafizrahman committed:
Avoid implicit init.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Use calculated constant instead of a var.

hafizrahman committed:
Declare coordinator lazily.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
weakly capture self in both outer and inner closures.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Remove unneeded initialization and simplify call.

hafizrahman committed:
Simplify check.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Remove unused function.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Simplification fix.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Make Product Form use the coordinator for Blaze too.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman commented:
@itsmeichigo 

Thanks for finding the regression! I missed the fact that multiple products case also exist in webview.

> Suggestion: You should inject an optional product ID to the coordinator. Then update the logic for creating the destination:

This is done in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/9f84b035a8632961e804653d90c6c7852dc97028

> Which brings me to the second point: you should integrate the coordinator to the product form too. The product ID in that case will always be present.

This is done in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/a8cbbb20fb183d8222a392d4b4a560a8ad7128f7

> Extra: it'd be nice to have unit tests for the coordinator to double check the logic. Please feel free to work on it in a separate PR.

Yes, this is going to be a separate PR as mentioned in the original PR post above.

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,173 @@
+
+    /// Handles navigation to the native Blaze creation
+    func navigateToNativeCampaignCreation(source: BlazeSource, productID: Int64) {
+        let viewModel = BlazeCampaignCreationFormViewModel(siteID: self.siteID,
```
```suggestion
        let viewModel = BlazeCampaignCreationFormViewModel(siteID: siteID,
```

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,173 @@
+    func navigateToNativeCampaignCreation(source: BlazeSource, productID: Int64) {
+        let viewModel = BlazeCampaignCreationFormViewModel(siteID: self.siteID,
+                                                           productID: productID,
+                                                           onCompletion: self.onCampaignCreated)
```
```suggestion
                                                           onCompletion: onCampaignCreated)
```

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,173 @@
+            let productSelectorViewModel = ProductSelectorViewModel(
+                siteID: siteID,
+                onProductSelectionStateChanged: { [weak self] product in
+                    guard let self = self else { return }
```
```suggestion
                    guard let self else { return }
```

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,173 @@
+                    navigateToNativeCampaignCreation(source: source, productID: product.productID)
+                },
+                onCloseButtonTapped: { [weak self] in
+                    guard let self = self else { return }
```
```suggestion
                    guard let self else { return }
```

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeCreationEntryCoordinator.swift:
```diff
@@ -0,0 +1,173 @@
+                    navigationController.dismiss(animated: true, completion: nil)
+                }
+            )
+            return ProductSelectorViewController(configuration: ProductSelectorView.Configuration.configurationForBlaze,
```
Nit: this can be simplified

```suggestion
            return ProductSelectorViewController(configuration: .configurationForBlaze,
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -1136,28 +1136,24 @@ private extension ProductFormViewController {
-            let webViewController = AuthenticatedWebViewController(viewModel: blazeViewModel)
-            navigationController?.show(webViewController, sender: self)
+    private func navigateToBlazeCampaignCreation(siteID: Int64, siteUrl: String, source: BlazeSource) {
+        guard let navigationController = self.navigationController else {
```
```suggestion
        guard let navigationController else {
            DDLogError("⛔️ Missing parent controller to show Blaze campaign creation form."
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -1136,28 +1136,24 @@ private extension ProductFormViewController {
+            navigationController: navigationController,
+            onCampaignCreated: handlePostCreation
+        )
+        coordinator.start()
```
We should keep a strong reference to the coordinator in the class just in case this local variable gets deallocated.

hafizrahman replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/11671/commits/fe0fd17eedb2a8a24806451cdc59a570e1a2d04e

itsmeichigo reviewed (changes requested):
Thanks for the updates, we're all good except for the missing of a strong reference to the coordinator in the `ProductFormViewController`. Missing that can lead to irresponsiveness in the Blaze campaign creation form if the coordinator gets deallocated.

hafizrahman committed:
remove self

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Also remove self

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/181">17.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Use shorthand.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
also use shorthand

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/181">17.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Simplify code.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Improve navigationController checking and usage.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/181">17.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Make strong reference to the blaze campaign coordinator.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/181">17.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo reviewed (approved):
All good now! 🚢  I'm merging this PR right away to use it in another PR.
